packages/context/src/binding-decorator.ts

```diff
@@ -0,0 +1,215 @@
+// Copyright IBM Corp. 2018. All Rights Reserved.
+// Node module: @loopback/context
+// This file is licensed under the MIT License.
+// License text available at https://opensource.org/licenses/MIT
+
+import {ClassDecoratorFactory, MetadataInspector} from '@loopback/metadata';
+import {Binding, BindingScope, BindingTemplate, TagMap} from './binding';
+import {Provider} from './provider';
+import {Constructor} from './value-promise';
+
+const BINDING_METADATA_KEY = 'binding.metadata';
+
+type BindingMetadata = {
+  templates: BindingTemplate[];
+  target?: Constructor<unknown>;
```

Comment:
What is the scenario where `target` ends up undefined? 

I think you need to define two interfaces:
 - configuration properties accepted by `@bind` decorator
 - the metadata object created by `@bind` decorator (config properties + target)